src/resolution.ts

```diff
@@ -0,0 +1,382 @@
+import type {
+  ParsedCaveatDefinition,
+  ParsedCaveatParameter,
+  ParsedExpression,
+  ParsedObjectDefinition,
+  ParsedPermission,
+  ParsedRelation,
+  ParsedRelationRefExpression,
+  ParsedSchema,
+  TopLevelDefinition,
+  TypeRef,
+  TextRange,
+} from "./dsl";
+import { flatMapExpression } from "./dsl";
+
+/**
+ * TypeRefResolution is the result of resolving a type reference.
+ */
+export interface TypeRefResolution {
+  /**
+   * definition is the definition which is referred by this type reference.
+   */
+  definition: ParsedObjectDefinition | undefined;
+
+  /**
+   * relation is the (optional) relation which is referred by this type reference. For refs
+   * without relations, this will be undefined. This will *also* be undefined for unresolvable
+   * refs, so make sure to check the expression to see if it has a relation reference.
+   */
+  relation: ParsedRelation | undefined;
+
+  /**
+   * permission is the (optional) permission which is referred by this type reference. For refs
+   * without relations, this will be undefined. This will *also* be undefined for unresolvable
+   * refs, so make sure to check the expression to see if it has a relation reference.
+   */
+  permission: ParsedPermission | undefined;
+}
+
+/**
+ * ExpressionResolution is the resolution of a reference in a permission expression.
+ */
+export type ExpressionResolution = ParsedRelation | ParsedPermission;
+
+/**
+ * ResolvedTypeReference is a type reference found in the schema, with resolution attempted.
+ */
+export interface ResolvedTypeReference {
+  kind: "type";
+  reference: TypeRef;
+  referencedTypeAndRelation: TypeRefResolution | undefined;
+}
+
+/**
+ * ResolvedExprReference is a relation reference expression found in the schema, with resolution attempted.
+ */
+export interface ResolvedExprReference {
+  kind: "expression";
+  reference: ParsedRelationRefExpression;
+  resolvedRelationOrPermission: ExpressionResolution | undefined;
+}
+
+/**
+ * ResolvedReference is a found and resolution performed type reference or expression.
+ */
+export type ResolvedReference = ResolvedTypeReference | ResolvedExprReference;
+
+/**
+ * Resolver is a helper class for easily resolving information in a parsed schema.
+ */
+export class Resolver {
+  private definitionsByName: Record<string, ResolvedDefinition> = {};
+  private caveatsByName: Record<string, ResolvedCaveatDefinition> = {};
+  private populated = false;
+  public schema: ParsedSchema = {
```

Comment:
I added default values here because the `private`/`public` declaration in the constructors fails the `noUnErasableSyntax` typescript rule, which we want because it makes working with TS easier.